### PR TITLE
feat: check if casks should warn about rosetta

### PR DESCRIPTION
**Pull request by SMillerDev**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----
This should warn in cases like https://github.com/Homebrew/homebrew-cask/pull/177913, so that authors can add the required caveats.

### CODE REVIEW

File: Library/Homebrew/cask/audit.rb [hunk 1]
```diff
@@ -562,6 +562,42 @@ def extract_artifacts
...
+    sig { void }
+    def audit_rosetta
+      return unless online?
+
+      odebug "Auditing rosetta requirement"
+
+      return if Homebrew::SimulateSystem.current_arch != :arm
+
+      extract_artifacts do |artifacts, tmpdir|
+        artifacts.filter { |a| a.is_a?(Artifact::App) || a.is_a?(Artifact::Binary) }.each do |artifact|
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/audit.rb:574]:
```suggestion
        artifacts.filter { |a| a.is_a?(Artifact::App) || a.is_a?(Artifact::Binary) }
                 .each do |artifact|
```

File: Library/Homebrew/cask/audit.rb [hunk 2]
```diff
@@ -562,6 +562,42 @@ def extract_artifacts
       yield artifacts, @tmpdir if block_given?
     end
 
+    sig { void }
+    def audit_rosetta
+      return unless online?
+
+      odebug "Auditing rosetta requirement"
+
+      return if Homebrew::SimulateSystem.current_arch != :arm
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/audit.rb:571]:
```suggestion
      return unless online?
      return if Homebrew::SimulateSystem.current_arch != :arm
      
      odebug "Auditing Rosetta 2 requirement"
```

File: Library/Homebrew/cask/audit.rb [hunk 3]
```diff
@@ -562,6 +562,42 @@ def extract_artifacts
...
+
+          unless result.success?
+            add_error "Failed to determine artifact architecture!", location: cask.url.location
+            next
+          end
+
+          next if result.merged_output.include?("arm64")
+          next if cask.caveats.include? "requires Rosetta 2"
+
+          add_error "Artifacts require Rosetta but this is not indicated by the caveats!", location: cask.url.location
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/audit.rb:596]:
```suggestion
          add_error "Artifacts require Rosetta 2 but this is not indicated by the caveats!", location: cask.url.location
```

**Review by MikeMcQuaid — APPROVED** [review]
Looks good! A few optional tweaks.

**Comment by SMillerDev** [comment]
Clearly I haven't been a mac user long enough, since I didn't mention Rosetta _2_ 😄.

Did another small addition, and included your changes @MikeMcQuaid

File: Library/Homebrew/cask/audit.rb [hunk 4]
```diff
@@ -562,6 +562,54 @@ def extract_artifacts
...
+          if !supports_arm && !mentions_rosetta
+            add_error "Artifacts require Rosetta 2 but this is not indicated by the caveats!",
+                      location: cask.url.location
+            next
+          end
+
+          if supports_arm && mentions_rosetta
+            add_error "Artifacts does not require Rosetta 2 but the caveats say otherwise!",
+                      location: cask.url.location
+          end
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/audit.rb:608]: Feel like this could probably be refactored to have a single `if`/`elsif`

**Comment by wickles** [comment]
This does not cover the 32-bit intel case from the referenced PR.

```
❯ brew audit --cask --online nightingale
==> Downloading and extracting artifacts
...
audit for nightingale: failed
 - Artifacts require Rosetta 2 but this is not indicated by the caveats!
nightingale
  * line 5, col 2: Artifacts require Rosetta 2 but this is not indicated by the caveats!
Error: 1 problem in 1 cask detected.
```

In this case rosetta caveat should not be added and another solution is needed instead as mentioned in the other PR.

**Comment by SMillerDev** [comment]
> This does not cover the 32-bit intel case from the referenced PR.

Those should be removed:
> Starting with macOS Catalina, 32-bit apps are no longer compatible with macOS.

https://support.apple.com/en-us/103076

**Comment by wickles** [comment]
> Those should be removed

Removed, disabled, or deprecated? `pgadmin3` in particular still gets lots of downloads (243 in 30 days).

**Pull request merged**